narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,233 +3,404 @@
         tokio::spawn(async move {
             Self {
+                state,
                 committee,
-                store,
+                certificate_store,
+                consensus_store,
                 rx_consensus_round_updates,
                 gc_depth,
                 rx_reconfigure,
-                rx_synchronizer,
-                tx_core,
-                pending: DashMap::new(),
-                metrics,
+                rx_certificate_waiter,
+                targets: BTreeMap::new(),
+                fetch_certificates_task,
             }
             .run()
             .await;
         })
     }
 
-    /// Helper function. It waits for particular data to become available in the storage and then
-    /// delivers the specified header.
-    async fn waiter(
-        missing: Vec<CertificateDigest>,
-        store: CertificateStore,
-        deliver: Certificate,
-        cancel_handle: oneshot::Receiver<()>,
-    ) -> DagResult<Certificate> {
-        let waiting: Vec<_> = missing.into_iter().map(|x| store.notify_read(x)).collect();
-
-        tokio::select! {
-            result = try_join_all(waiting) => {
-                result.map(|_| deliver).map_err(DagError::from)
-            }
-            // the request for this certificate is obsolete, for instance because its round is obsolete (GC'd).
-            _ = cancel_handle => Ok(deliver),
-        }
-    }
-
     async fn run(&mut self) {
-        let mut waiting = BoundedFuturesUnordered::with_capacity(self.max_pending_certificates());
-        let timer = sleep(Duration::from_millis(GC_RESOLUTION));
-        tokio::pin!(timer);
-        let mut attempt_garbage_collection;
-
-        info!("CertificateWaiter has started successfully.");
         loop {
-            // Initially set to not garbage collect
-            attempt_garbage_collection = false;
-
             tokio::select! {
-                // We only accept new elements if we have "room" for them
-                Some(certificate) = self.rx_synchronizer.recv(), if waiting.available_permits() > 0 => {
-                    if certificate.epoch() < self.committee.epoch() {
+                Some(certificate) = self.rx_certificate_waiter.recv() => {
+                    let header = &certificate.header;
+                    if header.epoch != self.committee.epoch() {
                         continue;
                     }
+                    // Unnecessary to validate the header and certificate further, since it has
+                    // already been validated.
 
-                    // Ensure we process only once this certificate.
-                    let header_id = certificate.header.id;
-                    if self.pending.contains_key(&header_id) {
-                        continue;
+                    if let Some(r) = self.targets.get(&header.author) {
+                        if header.round <= *r {
+                            // Ignore fetch request when we already need to sync to a later
```

Comment:
Yes this is a bit subtle and I'm adding more comments. The basic reason is that it is ok to stop fetching even when a validator is missing some certificates, because the validator will kick off another fetch once it sees another missing certificate.

At a high level, certificate fetching requires three decisions:
1. When to start fetching certificates (trigger): seeing a certificate missing parents seems like a good choice. Or a validator can always issue a fetch request every 5s.
2. What certificates to fetch:
   - Always starting from round 0 is safe but inefficient. 
   - Starting from the last written round of each origin can be incorrect.
   - Causal completion takes many round trips to figure out the missing certificates.
   - This implementation starts from last committed rounds. In future we will also add treeset / bitmap to represent certificates between last committed and last written rounds. This way, a missing certificate will either be above last committed round so it will get fetched eventually, or below last committed round so fetching it is unnecessary.
3. When to stop fetching:
   - issuing fetch request every 5s forever may be reasonable, although adding cost to the system.
   - Keeping track of all received certificates with missing parents are also reasonable, as long as the total size is bounded.
   - This implementation stops fetching after the highest seen round of each origin is received. As you described, this may terminate the fetching too early. But as long as another missing certificate is received, a new fetch will start.

For the implementation here, when a validator A discovers it needs Cb, it will try to fetch certificates for all origins starting from last committed rounds, from all validators, until Cb is committed. e.g.
1. Validator A discovers it is missing Cb at round 101. It sends a fetch request containing its last committed rounds for all validators, e.g. round 98 for all validators `Vx = Vy = Vz = Vq = 98`.
2. A honest validator B containing Cb and receiving the fetch request will notice A has not committed any certificate at round 99 or 100, so will send back all its certificates at these rounds including Cq and Cb.
3. A dishonest validator or a validator missing Cb / Cq may not help A get Cb, but A will keep running the fetch task.
5. If somehow A does not have Cq but commits a higher round certificate for validator q, e.g. at round 102, validator q round 100 will not be in the targets. But this is fine because Cq will not make into the dag in this case anyway.

The stop fetch criteria is a balance between memory usage for tracking missing certificates, cost of issuing and processing fetch requests, and latency increase for processing certificates when they are missing parents. I'm open to more discussions if you want to.